### ISSUE TITLE
[VL] Vanilla Spark broadcast exchange + R2C is slow sometimes

### ISSUE THREAD

**Issue by zhztheplayer** [apache/incubator-gluten#5136]
### Backend

VL (Velox)

### Bug description

This is because the code to convert vanilla Spark's hashed relation to Gluten's sometimes produced duplicated rows.

The fix will be incorporated in https://github.com/apache/incubator-gluten/pull/5058 (at commit https://github.com/apache/incubator-gluten/pull/5058/commits/96c3fc7b282370e49012bcfff17ab243f2bdc33f) since it can be tested by the ACBO changes.

**Comment by ulysses-you** [comment]
Thank you @zhztheplayer It's a good point, columnar broadcast would broadcast the origin binary data but vanilla Spark would broadcast hash relation. So I think this issue is a common case even if there is no r2c.

Is it possbile to create a new pr for this issue ?

**Comment by zhztheplayer** [comment]
I don't have dedicated UTs for it so it was incorporated into the other PR.

Still I can open one for it if you think it's needed: https://github.com/apache/incubator-gluten/pull/5141.

The change was already tested so I will proceed to merge after code style check is passed if it's OK to you.

**Comment by zhztheplayer** [comment]
The major issue I have found is that the `flatMap` approach would cause `UnsafeHashedRelation` to produce duplicated rows in my case (TPCDS q14a with current version of ACBO)
While the `map` approach would cause `LongHashedRelation` to loss rows (TPCDS q2).

The following fix (the same with #5141) can work but I didn't dive into it deeply to find the root reason of the inconsistency (maybe related to `keyIsUnique`? I am not sure).

```scala
  private def reconstructRows(relation: HashedRelation): Iterator[InternalRow] = {
    // It seems that LongHashedRelation and UnsafeHashedRelation don't follow the same
    //  criteria while getting values from them.
    // Should review the internals of this part of code.
    relation match {
      case relation: LongHashedRelation if relation.keyIsUnique =>
        relation.keys().map(k => relation.getValue(k))
      case relation: LongHashedRelation if !relation.keyIsUnique =>
        relation.keys().flatMap(k => relation.get(k))
      case other => other.valuesWithKeyIndex().map(_.getValue)
    }
  }
```

**Comment by zhztheplayer** [comment]
Fixed in https://github.com/apache/incubator-gluten/pull/5141. I assume we can close this now.
